### PR TITLE
deprecate the 'empty_array_mt' field in favor of 'array_mt'

### PR DESCRIPTION
**Pull request by thibaultcha**
As discussed in #2, this deprecates the `cjson.empty_array_mt` field in favor of `cjson.array_mt`. The new field has the same behavior as the previous one, but simply bears a more explicit name.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -80,6 +81,16 @@ empty_array_mt
 --------------
 **syntax:** `setmetatable({}, cjson.empty_array_mt)`
 
+**NOTE**: Use of this field is *discouraged* as of the `2.1.0.6` release. This
```
Comment (thibaultcha) [inline README.md]: This version number is a guess based on the current release versions, to be updated if necessary.

File: README.md [hunk 2]
```diff
@@ -80,6 +81,16 @@ empty_array_mt
...
+field is considered *depcrecated*, and has been renamed to
+[array_mt](#array_mt) for expliciteness.
+
+[Back to TOC](#table-of-contents)
+
+array_mt
+--------
+**syntax:** `setmetatable({}, cjson.array_mt)`
+
 A metatable which can "tag" a table as a JSON Array in case it is empty (that is, if the
```
Comment (agentzh) [inline README.md]: This text is still confusing. It shouldn't matter if the table is empty or not. It won't even matter when the table actually has a hash part.

File: tests/agentzh.t [hunk 3]
```diff
@@ -113,7 +113,26 @@ print(cjson.encode(data))
...
+
+
+=== TEST 10: array_mt behaves as empty_array_mt
+--- lua
+local cjson = require "cjson"
+local empty_arr = setmetatable({}, cjson.array_mt)
+print(cjson.encode({obj = {}, arr = empty_arr}))
+--- out
+{"arr":[],"obj":{}}
+
```
Comment (agentzh) [inline tests/agentzh.t]: Need to test non-empty tables with `cjson.array_mt` as well, including those use the hash table to store the array elements, as in

```
local new_tab = require "table.new"
local t = new_tab(0, 2)  -- only allocate the hash part
t[1] = "hello"
setmetatable(t, cjson.array_mt)
cjson.encode(t)  -- should be ["hello"] here
```

File: README.md (hunk 2 above)
Comment (thibaultcha) [inline README.md]: So the behavior we want from `cjson.array_mt` is slightly different than `cjson.empty_array_mt` then? Currently, `cjson.empty_array_mt` encodes a table as an Array if it is empty (avoids having to do the `if #t < 1` thing.

What it seems we want here is something that encodes as a JSON Array regardless of its content? It then seems both metatables belong to this module as they solve different use cases?

Comment (agentzh) [inline README.md]: @thibaultcha The whole idea is to avoid the user to call `#tb` to test the array length but simply plants `cjson.array_mt` into all the array tables.

Comment (thibaultcha) [inline README.md]: Yes, that's what `cjson.empty_array_mt` does already; it avoids calling `#t` already. So I'm very unsure what the difference is here.

Comment (agentzh) [inline README.md]: @thibaultcha If there is no difference, then why the document above still specifically talks about "empty"?

Comment (agentzh) [inline README.md]: @thibaultcha The user of cjson may not know if the tables are empty or not without testing for themselves. So requiring "empty" in the documentation is asking for such test on the user side.

Comment (thibaultcha) [inline README.md]: Maybe the documentation wording is not clear enough. What cjson does about it is here:

https://github.com/openresty/lua-cjson/blob/master/lua_cjson.c#L716-L724

As we can see from that logic, it deals with ambiguities when a table has a 0 length:

  * if cjson is encoding a table:
    * that table has `len > 0` -> array `[...]`
    * table has `len == 0`
      * if `!cfg.encode_empty_table_as_object` -> array `[]`
      * if table's metatable is the `cjson_empty_array` -> array `[]`
      * else -> object `{}`

The length of the table comes from [this function](https://github.com/openresty/lua-cjson/blob/master/lua_cjson.c#L520) (link reference only). It seems to simply rely on the presence of non-integer keys yielded by `lua_next()` to determine if a table is an Object/Array.

What I think I understand from your use case, is that is it about having another metatable, one that if set on a table that has both a hash and an array part (it would fail the current check from `lua_array_length` because it has non-integer keys), then still *force* the encoding of the array part only, as a table. Is that correct?

Comment (agentzh) [inline README.md]: @thibaultcha Yes, that's correct.

Comment (thibaultcha) [inline README.md]: @agentzh Ok, thanks, that's clearer to me know. I'll take a stab at it! Sorry I misunderstood the use-case in this PR.

Comment (agentzh) [inline README.md]: @thibaultcha Awesome! Thank you so much!

Another thing worth considering is to add an option to make `decode` automatically set `array_mt` on empty arrays :)

Comment (agentzh) [inline README.md]: @thibaultcha Or all arrays.

Comment (agentzh) [inline README.md]: @thibaultcha Also, it would be nice if you can benchmark the performance before and after your changes too :)

Comment (thibaultcha) [inline README.md]: Sounds good, I'll take some time tomorrow or this week to look into this. I'll be closing this PR in the meantime.
